src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsCoreProjectSystemReferenceReader.cs

```diff
@@ -124,7 +136,26 @@ internal class VsCoreProjectSystemReferenceReader
                 logger.LogVerbose(message);
             }
 
+            if (hasProjectsWithUnresolvedMetadata)
+            {
+                IList<string> excludedProjects = await GetExcludedProjectsAsync(logger);
+                if (excludedProjects.Count > 0)
+                {
+                    results = results.Where(e => excludedProjects.Contains(e.ProjectPath, StringComparer.OrdinalIgnoreCase)).ToList();
```

Comment:
Callers to this method are invoking `.ToList()` method explicitly (couple of examples are listed below). May be invoking `.ToList()` here is not required. It is possible I am missing something here as I am looking at this code path for the first time. 

For e.g. https://github.com/NuGet/NuGet.Client/blob/d2442f173798ae9f2d95cda34cc156b4517afcdf/src/NuGet.Core/NuGet.PackageManagement/Projects/MSBuildNuGetProject.cs#L687-L688

https://github.com/NuGet/NuGet.Client/blob/f24bad0668193ce21a1db8cabd1ce95ba509c7f0/src/NuGet.Core/NuGet.PackageManagement/Projects/ProjectJsonNuGetProject.cs#L223-L224

Reply:
I get your point, but I'm trying to focus on the implementation of the current method here. 

I think it'd be better if the calling methods would call `AsList` instead, but ideally that's done in a different PR.

Reply:
Isn't it `results = results.Where(e => !excludedProjects.Contains(e.ProjectPath, StringComparer.OrdinalIgnoreCase)).ToList();` based on context? Exclude mean we don't want, right? I'm trying to grasp why only project in excluded are now returned, correct me if I'm wrong.

Reply:
If above is the case maybe below is more concise. It looks `Except` uses [HashSet internally](http://index/?leftProject=System.Core&leftSymbol=ktweftexyual&file=System%5CLinq%5CEnumerable.cs&line=899) could be more performant on large list, but could be more overhead on if 2nd list is small.
![image](https://user-images.githubusercontent.com/8766776/141228134-7e4bb889-7770-4648-a88b-3fe3a2aa361b.png)

Reply:
I think the perf concerns are not super important cause `ReferenceOutputAssembly` is not frequently used, and those list are always small. 

I'll take a look at the condition.

Reply:
Fixed. 
There isn't really a test case that can actually repro that reliably.

ServiceFabric projects are supposed to be it, but the refs are not resolved there.